src/search.cpp

```diff
@@ -433,9 +433,9 @@ void Thread::search() {
           if (rootDepth >= 4)
           {
               Value previousScore = rootMoves[pvIdx].previousScore;
-              delta = Value(21 + abs(previousScore) / 256);
+              delta = 21 + abs(previousScore) / 256;
               alpha = std::max(previousScore - delta,-VALUE_INFINITE);
-              beta  = std::min(previousScore + delta, VALUE_INFINITE);
+              beta  = std::min(previousScore + delta, Value(VALUE_INFINITE));
```

Comment:
@protonspring I still think we need to understand why this makes a difference. The type of VALUE_INFINITE is Value, so this should not matter. Does the compiler really generate a different binary with and without this change?

Reply:
Yes. If I don't type cast VALUE_INFINITE the resulting binary increases in size and will no longer pass -3,1 (I tried 10 times over the past 6 months).  It's stupid and makes no sense to me.

Reply:
I can reproduce this, but it depends on the compiler version. With g++-7 I have indeed a difference in generated code, while with g++-8 the code is identical, as it should be.

This section of the code should also be uncritical from the performance point of view, so it must be something that trickles down to some more important place... strange.

Reply:
This might be a missing optimization in g++7, in case GCC considers enum values as constexpr variables. (That I don't know for a fact.)

std::min() prototype is:
<code>template< class T >
constexpr const T& min( const T& a, const T& b );</code>

Const lvalue references accept either references to objects or references to temporary objects. In principle, it is at least possible that VALUE_INFINITE is materialized as an actual constexpr variable, and an actual reference is passed to std::min here. Passing a reference would arguably be slightly slower than using immediate value, since occasional L1 or L2 cache misses would be expected.

However, if an expression is passed as an argument (e.g., VALUE(VALUE_INFINITE) or -VALUE_INFINITE or similar), then reference to real variable cannot be used. Rather, reference to temporary object would be used and that is likely to be inlined in the template function instantiation.

I don't have GCC-7 handy so I can't inspect the disassembly to see whether the above is the case. But GCC in my experience likes to take actual references to constexpr variables with std::min/max, so at least this is a candidate explanation.

Reply:
Note: Sometimes the GCC eagerness of taking references to constexprs will cause linker errors due to missing symbols to those references, since constexpr implies inline (or something like that). Was that the reason why you added the Value(VALUE_INFINITE) cast here?

Reply:
I understand very little here.  I only added the cast because it prevented an unexpected change to the binary and subsequent failure to pass STC -3,1.  When I add the cast, there is no slowdown and passes -3,1 easily.

Reply:
I understand very little here. I only added the cast because it prevented an unexpected change to the binary and subsequent failure to pass STC -3,1. When I add the cast, there is no slowdown and passes -3,1 easily.